rust-mozjs/examples/wasm.rs

```diff
@@ -0,0 +1,179 @@
+
+    // Get WebAssembly.Module and WebAssembly.Instance constructors.
+    rooted!(in(rt.cx()) let mut wasm = UndefinedValue());
+    rooted!(in(rt.cx()) let mut wasm_module = UndefinedValue());
+    rooted!(in(rt.cx()) let mut wasm_instance = UndefinedValue());
+
+    unsafe {
+        assert!(JS_GetProperty(
+            rt.cx(),
+            global.handle(),
+            b"WebAssembly\0".as_ptr() as *const libc::c_char,
+            &mut wasm.handle_mut()
+        ));
+        rooted!(in(rt.cx()) let mut wasm_obj = wasm.to_object());
+        assert!(JS_GetProperty(
+            rt.cx(),
+            wasm_obj.handle(),
+            b"Module\0".as_ptr() as *const libc::c_char,
+            &mut wasm_module.handle_mut()
+        ));
+        assert!(JS_GetProperty(
+            rt.cx(),
+            wasm_obj.handle(),
+            b"Instance\0".as_ptr() as *const libc::c_char,
+            &mut wasm_instance.handle_mut()
+        ));
+
+        // Construct Wasm module from bytes.
+        rooted!(in(rt.cx()) let mut module = null_mut::<JSObject>());
+        {
+            let array_buffer = JS::NewArrayBufferWithUserOwnedContents(
+                rt.cx(),
+                HI_WASM.len(),
+                HI_WASM.as_ptr() as _,
+            );
+            assert!(!array_buffer.is_null());
+
+            rooted!(in(rt.cx()) let val = ObjectValue(array_buffer));
+            let args = HandleValueArray {
+                length_: 1,
+                elements_: &*val,
+            };
+
+            assert!(Construct1(
+                rt.cx(),
+                wasm_module.handle(),
+                &args,
+                &mut module.handle_mut()
+            ))
+        }
+
+        // Construct Wasm module instance with required imports.
+        rooted!(in(rt.cx()) let mut instance = null_mut::<JSObject>());
+        {
+            // Build "env" imports object.
+            rooted!(in(rt.cx()) let mut env_import_obj = JS_NewPlainObject(rt.cx()));
+            assert!(!env_import_obj.is_null());
+            let function = JS_DefineFunction(
+                rt.cx(),
+                env_import_obj.handle().into(),
+                b"bar\0".as_ptr() as *const libc::c_char,
+                Some(bar),
+                1,
+                0,
+            );
+            assert!(!function.is_null());
+            rooted!(in(rt.cx()) let mut env_import = ObjectValue(env_import_obj.get()));
+            // Build imports bag.
+            rooted!(in(rt.cx()) let mut imports = JS_NewPlainObject(rt.cx()));
+            assert!(!imports.is_null());
+            assert!(JS_SetProperty(
+                rt.cx(),
+                imports.handle(),
+                b"env\0".as_ptr() as *const libc::c_char,
+                env_import.handle()
+            ));
+
+            let args_slice = [ObjectValue(module.get()), ObjectValue(imports.get())];
+            let args = HandleValueArray::from_rooted_slice(&args_slice);
```

Comment:
I am not 100% sure about this one, I did it like it is done in [servo/components/script/dom/paintworkletglobalscope.rs](https://github.com/servo/servo/blob/b1d90f871dd4ddddbd6f4fa66648265674338e29/components/script/dom/paintworkletglobalscope.rs#L344).

Reply:
That code needs to be fixed as well ;_;

Reply:
Not a Rust unsafe experienced but wrapping `[ObjectValue(module.get()), ObjectValue(imports.get())]` in `ManuallyDrop` should be enough?

Reply:
The safety risk here is that the garbage collector does not know about the slice that contains those JS values, so if they are moved for any reason like a GC occurs during the next spidermonkey API call), the JS values will not be updated with the new pointers. This means the handles passed as arguments would be invalid and could cause a crash.

Reply:
I filed https://github.com/servo/mozjs/issues/346 about the missing pieces to make this safe.

Reply:
Hm, but we do have `CustomTrace` and `auto_root!` so 
```rust
auto_root!(in(rt.cx()) let vec = vec![ObjectValue(module.get()), ObjectValue(imports.get())]);
```
should do it, right?